### PR TITLE
Show not keyword (and all other stuff) after and & or when they are treated as identifiers

### PR DESCRIPTION
**Pull request by DoctorKrolic**
Fixes: https://github.com/dotnet/roslyn/issues/61184

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs [hunk 1]
```diff
@@ -1383,6 +1383,15 @@ public static bool IsAtStartOfPattern(this SyntaxTree syntaxTree, SyntaxToken le
                 return leftToken.Parent is BinaryPatternSyntax;
             }
 
+            // e is SomeEnum.SomeEnumValue and $$
+            // e is SomeEnum.SomeEnumValue or $$
+            // 'and' & 'or' are identifier here because of lack of context
+            if (leftToken.IsKind(SyntaxKind.IdentifierToken) &&
+                (leftToken.IsKindOrHasMatchingText(SyntaxKind.AndKeyword) || leftToken.IsKindOrHasMatchingText(SyntaxKind.OrKeyword)))
```
Comment (Youssef1313) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs:1390]: What happens if this became just this:

```suggestion
            if (leftToken.IsKindOrHasMatchingText(SyntaxKind.AndKeyword) || leftToken.IsKindOrHasMatchingText(SyntaxKind.OrKeyword))
```

Anything will break? If yes, consider making sure there is a test that will break if this change is made.

**Comment by DoctorKrolic** [comment]
@CyrusNajmabadi Test with `not` passed, so I included it. Test with `and` didn't. I feel like it is a separate issue and it should be fixed in other PR

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs [hunk 2]
```diff
@@ -1378,9 +1378,12 @@ public static bool IsAtStartOfPattern(this SyntaxTree syntaxTree, SyntaxToken le
 
             // e is 1 and $$
             // e is 1 or $$
-            if (leftToken.IsKind(SyntaxKind.AndKeyword) || leftToken.IsKind(SyntaxKind.OrKeyword))
+            // e is SomeEnum.SomeEnumValue and $$
+            // e is SomeEnum.SomeEnumValue or $$
+            // 'and' & 'or' are identifier in the last 2 examples because of lack of context
+            if (leftToken.IsKindOrHasMatchingText(SyntaxKind.AndKeyword) || leftToken.IsKindOrHasMatchingText(SyntaxKind.OrKeyword))
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs:1384]: ❓ Shouldn't this include `leftToken.IsKindOrHasMatchingText(SyntaxKind.NotKeyword)`? Patterns are not not allowed to use this.

File: src/EditorFeatures/CSharpTest2/Recommendations/NotKeywordRecommenderTests.cs [hunk 3]
```diff
@@ -285,5 +293,25 @@ public async Task TestAfterRelationalOperator()
...
+
+        [WorkItem(61184, "https://github.com/dotnet/roslyn/issues/61184")]
+        [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
+        public async Task TestAfterIdentifierAnd()
+        {
+            await VerifyKeywordAsync(InitializeObjectE +
+@"if (e is Test.TestValue and $$)
+
+enum Test { TestValue }");
+        }
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest2/Recommendations/NotKeywordRecommenderTests.cs:305]: 💡 A `Theory` here eliminates the need for multiple tests that have the same basic form.

```suggestion
        [WorkItem(61184, "https://github.com/dotnet/roslyn/issues/61184")]
        [Theory, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
        [InlineData("and")]
        [InlineData("or")]
        [InlineData("not")]
        public async Task TestAfterIdentifierAnd(string precedingKeyword)
        {
            await VerifyKeywordAsync(InitializeObjectE +
$@"if (e is Test.TestValue {precedingKeyword} $$)

enum Test {{ TestValue }}");
        }
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs [hunk 4]
```diff
@@ -1378,9 +1378,12 @@ public static bool IsAtStartOfPattern(this SyntaxTree syntaxTree, SyntaxToken le
...
             {
-                return leftToken.Parent is BinaryPatternSyntax;
+                return leftToken.Parent is BinaryPatternSyntax || leftToken.Parent?.Parent is DeclarationPatternSyntax;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs:1386]:
```suggestion
                return leftToken.Parent is BinaryPatternSyntax || leftToken.Parent is SingleVariableDesignationSyntax { Parent: DeclarationPatternSyntax };
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs (hunk 2 above)
Comment (DoctorKrolic) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs:1384]: Can you please give an example where `not not` is a valid pattern. I cannot understand how it is possible

**Comment by CyrusNajmabadi** [comment]
>  I feel like it is a separate issue and it should be fixed in other PR

Are you ok if I push to this branch @DoctorKrolic ?

**Comment by DoctorKrolic** [comment]
@CyrusNajmabadi Ok. Still don't understand, why you want it to be fixed here anyway...

**Comment by CyrusNajmabadi** [comment]
I'd prefer to take care of these issues all in one go since we're looking in this space anyways. It's more efficient that way vs having to come back at some random point in the future and page all this code in.  Thanks!

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs (hunk 2 above)
Comment (Neme12) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs:1384]: @sharwell `not` is always valid after `not`. This code is valid:
```c#
void M(int a)
{
    if (a is not not 7)
    {
    }
}
```

Comment (DoctorKrolic) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs:1384]: @Neme12 Didn't know that. However, in our case second not will not form a valid pattern:
```cs
var a = DayOfWeek.Monday;

if (a is DayOfWeek.Saturday not not DayOfWeek.Monday)
{

}
```
This code produces 9 errors.

File: src/EditorFeatures/CSharpTest2/Recommendations/AndKeywordRecommenderTests.cs [hunk 5]
```diff
@@ -553,5 +551,25 @@ public enum Goo
...
+        [WorkItem(51431, "https://github.com/dotnet/roslyn/issues/51431")]
+        [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
+        public async Task TestAfterParenthesizedPattern1()
+        {
+            await VerifyKeywordAsync(
+@"
+expr is (not []) $$");
+        }
+
+        [WorkItem(51431, "https://github.com/dotnet/roslyn/issues/51431")]
```
Comment (DoctorKrolic) [inline src/EditorFeatures/CSharpTest2/Recommendations/AndKeywordRecommenderTests.cs:564]: @CyrusNajmabadi Is this `WorkItem` right? I am not sure. The same in the test above

**Pull request merged**